rs/nervous_system/tools/sync-with-released-nervous-system-wasms/src/main.rs

```diff
@@ -79,68 +100,100 @@ async fn get_mainnet_canister_git_commit_id_and_module_hash(
+
+        let asset_bytes = client
+            .get(self.browser_download_url.clone())
+            .bearer_auth(&token)
+            .send()
+            .await?
+            .bytes()
+            .await?;
+
+        Ok(module_hash_hex(sha2::Sha256::digest(&asset_bytes).to_vec()))
     }
 
-    const GITHUB_API: &str = "https://api.github.com";
+    // Returns the textual content of the asset.
+    async fn text(&self) -> Result<String> {
+        let (client, token) = github_api_client_and_token()?;
 
-    let canister_id = canister_id.get().0;
+        Ok(client
+            .get(self.browser_download_url.clone())
+            .bearer_auth(&token)
+            .send()
+            .await?
+            .text()
+            .await?)
+    }
+}
 
-    let module_hash = agent
-        .read_state_canister_info(canister_id, "module_hash")
-        .await?;
-    let module_hash = module_hash.iter().fold(String::new(), |mut output, x| {
-        let _ = write!(output, "{:02x}", x);
-        output
-    });
+#[derive(Debug, Deserialize)]
+struct Release {
+    tag_name: String,
+    assets: Vec<ReleaseAsset>,
+}
 
-    let token = std::env::var("GITHUB_TOKEN").expect("Set GITHUB_TOKEN env var");
+impl Release {
+    // Finds and returns a release asset based on its filename
+    // if the asset is part of the release.
+    fn asset(&self, filename: &str) -> Option<ReleaseAsset> {
+        self.assets
+            .iter()
+            .find(|asset| asset.name == *filename)
+            .cloned()
+    }
+}
 
-    // GitHub API requires all requests to provide a user-agent header
-    // so we fix an arbitrary value of this header here.
-    let client = Client::builder()
-        .user_agent("sync-with-released-nervous-system-wasms")
-        .build()?;
+/// This function finds a release in the given `canister_repository` (e.g., `dfinity/cycles-ledger`)
+/// which contains a release asset for the given canister with the expected sha256 hash.
+/// To find the release, this function proceeds as follows:
+///   - it crawls all git tags of the given `canister_repository`;
+///   - for every git tag, it checks if there is an associated release and then
+///     - looks for a release asset whose name has the form `{canister_name}.sha256`:
+///       if that release asset starts with `expected_module_hash_str`, then the corresponding release is returned;
+///     - otherwise, this function looks for a release asset whose name matches `canister_name`:
+///       if the sha256 hash of that release asset matches `expected_module_hash_str`, then the corresponding release is returned.
+async fn get_mainnet_canister_release(
+    canister_name: String,
+    canister_repository: String,
+    canister_filename: String,
+    expected_module_hash_str: String,
+) -> Result<Release> {
+    let (client, token) = github_api_client_and_token()?;
 
     let tags_url = format!("{}/repos/{}/tags", GITHUB_API, canister_repository);
     let tags: Vec<Tag> = client
```

Comment:
Do we know in what order do these get listed? Are they never paginated (I assume some projects can have 10K-100K tags)?

Reply:
They are in reverse chronological order, but they are indeed paginated (good catch!). I'll make the crawling more robust.

Reply:
Apparently, the external canisters added so far are hit on the first page.

Reply:
I've added pagination now: https://github.com/dfinity/ic/pull/6010/commits/7d9d9ededc1837fff3bc73b5ec559f2795737b7e